mac/Keyman4MacIM/Keyman4MacIM/KMDataRepository.m

```diff
@@ -0,0 +1,185 @@
+@property (readonly) NSURL *documentsSubDirectory;    // '~/Documents'
+@property (readonly) NSURL *obsoleteKeymanKeyboardsDirectory; // '~/Library/Documents/Keyman-Keyboards'
+@end
+
+@implementation KMDataRepository
+
+@synthesize applicationSupportSubDirectory = _applicationSupportSubDirectory;
+@synthesize documentsSubDirectory = _documentsSubDirectory;
+@synthesize keymanKeyboardsDirectory = _keymanKeyboardsDirectory;
+@synthesize obsoleteKeymanKeyboardsDirectory = _obsoleteKeymanKeyboardsDirectory;
+@synthesize keymanDataDirectory = _keymanDataDirectory;
+
+NSString *const kKeyboardsDirectoryName = @"Keyman-Keyboards";
+/* 
+ The name of the subdirectory within '~/Library/Application Support'.
+ The convention is to use bundle identifier.
+ (Using the subsystem id, "com.keyman.app", is a poor choice because the API
+ createDirectoryAtPath sees the .app extension and creates an application file.
+ */
+NSString *const kKeymanSubdirectoryName = @"keyman.inputmethod.Keyman";
+
++ (KMDataRepository *)shared
+{
+  static KMDataRepository *shared = nil;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    shared = [[KMDataRepository alloc] init];
+  });
+  return shared;
+}
+
+- (NSURL *)documentsSubDirectory {
+  if (_documentsSubDirectory == nil) {
+    NSError *directoryError = nil;
+    
+    NSFileManager *fileManager = [NSFileManager defaultManager];
+    NSURL *documentsUrl = [fileManager URLForDirectory:NSDocumentDirectory inDomain:NSUserDomainMask appropriateForURL:nil create:YES error:&directoryError];
+    
+    if (directoryError) {
+      os_log_error([KMLogs dataLog], "error getting Documents subdirectory: '%{public}@'", directoryError.localizedDescription);
+    } else {
+      os_log_info([KMLogs dataLog], "Documents subdirectory: '%{public}@'", documentsUrl);
+      _documentsSubDirectory = documentsUrl;
+    }
+  }
+  return _documentsSubDirectory;
+}
+
+- (NSURL *)applicationSupportSubDirectory {
+  if (_applicationSupportSubDirectory == nil) {
+    NSError *directoryError = nil;
+    
+    NSFileManager *fileManager = [NSFileManager defaultManager];
+    NSURL *applicationSupportUrl = [fileManager URLForDirectory:NSApplicationSupportDirectory inDomain:NSUserDomainMask appropriateForURL:nil create:YES error:&directoryError];
+
+    if (directoryError) {
+      os_log_error([KMLogs dataLog], "error getting Application Support subdirectory: '%{public}@'", directoryError.localizedDescription);
+    } else {
+      os_log_info([KMLogs dataLog], "Application Support subdirectory: '%{public}@'", applicationSupportUrl.path);
+      _applicationSupportSubDirectory = applicationSupportUrl;
+    }
+  }
+  return _applicationSupportSubDirectory;
+}
+
+- (NSURL *)keymanDataDirectory {
+  if (_keymanDataDirectory == nil) {
+    NSURL *keymanDataUrl = [self.applicationSupportSubDirectory URLByAppendingPathComponent:kKeymanSubdirectoryName  isDirectory: TRUE];
```

Comment:
I cannot think of anything intelligent to do other than log an error if the OS does not return it to me. This is the place that we are supposed to create our app subdirectory for our user data, so if it doesn't exist, then we don't have a fallback location that I am aware of. In my 'Application Support' directory, there are 91 subdirectories created by other software, so it would be a major failure if it weren't there.